### ISSUE TITLE
FSGetVolumeParms() returns 0 for bSupports2TBFiles on FUSE volumes

### ISSUE THREAD

**Issue by swt2c** [osxfuse/osxfuse#373]
When calling `FSGetVolumeParms()` on a FUSE volume, the resulting value for bSupports2TBFiles, which is supposed to indicate support for > 4GB-sized files, is 0.

Based on some old mailing list traffic, Finder was at one point calling this function, which is now deprecated.  However, it may still be calling it:
https://lists.apple.com/archives/darwin-kernel/2007/Nov/msg00092.html

Sample code:
```
#import <Foundation/Foundation.h>

int main(int argc, char *argv[])
{
  OSStatus result = noErr;
  ItemCount volumeIndex;
  for (volumeIndex = 1; result == noErr || result != nsvErr; volumeIndex++) {
    FSVolumeRefNum actualVolume;
    HFSUniStr255 volumeName;
    FSVolumeInfo volumeInfo;

    bzero((void *) &volumeInfo, sizeof(volumeInfo));

    result = FSGetVolumeInfo(kFSInvalidVolumeRefNum,
			     volumeIndex,
			     &actualVolume,
			     kFSVolInfoFSInfo,
			     &volumeInfo,
			     &volumeName,
			     NULL);

    if (result == noErr ) {
      GetVolParmsInfoBuffer volumeParams;
      result = FSGetVolumeParms(actualVolume, &volumeParams, sizeof(volumeParams));
      CFStringRef stringRef = FSCreateStringFromHFSUniStr(nil, &volumeName);
      NSLog(@"BLAH: %@ %d", stringRef, volumeParams.vMExtendedAttributes & bSupports2TBFiles);
    }
  }
}
```

**Comment by bfleischer** [comment]
I'm able to reproduce the issue, but I don't know why `FSGetVolumeParms()` doesn't set `bSupports2TBFiles`. `VOL_CAP_FMT_2TB_FILESIZE` is clearly set for the volume. The algorithm outlined in the liked post does not seem to match the behavior I'm seeing on macOS Sierra.

**Comment by swt2c** [comment]
I don't know why it is behaving that way, either.  My hunch is that it somehow isn't able to call `getattrlist()` successfully at mount time.

**Comment by bfleischer** [comment]
Ok, it looks like we need to wait until Apple fixes the issue. Please let me know if your radar turns up any information how we can work around this. Without knowing more about the internal workings of corservicesd there is not much I can do.

**Comment by bfleischer** [comment]
`bSupports2TBFiles` is the bit number, the actual bit mask for checking if the volume supports files greater than 4 GB is `(1L << bSupports2TBFiles)`. I didn't catch that either.

Can you confirm that the `(1L << bSupports2TBFiles)` bit is set for your FUSE file system?

**Comment by swt2c** [comment]
Yes, Jim pointed out that error, too.

After fixing the code to check correctly, it seems that `bSupports2TBFiles` IS set for FUSE filesystems on most systems.  However, on the two systems where I am seeing errors from Finder, `bSupports2TBFiles` IS NOT set.  Thus, the question remains - why is `FSGetVolumeParms()` not returning the correct answer on these two systems?

I let Jim know about this, but I haven't heard anything further from him yet.

**Comment by swt2c** [comment]
One other data point...if I run the `FSGetVolumeParms()` check as root on those two systems, it reports `bSupports2TBFiles` as SET.

**Comment by bfleischer** [comment]
> However, on the two systems where I am seeing errors from Finder, bSupports2TBFiles IS NOT set.

That's interesting. I don't think this is coincidence. 

> One other data point...if I run the FSGetVolumeParms() check as root on those two systems, it reports bSupports2TBFiles as SET.

Are you mounting your file system with `-o allow_other`?

**Comment by swt2c** [comment]
> Are you mounting your file system with -o allow_other?

No.

**Comment by bfleischer** [comment]
Could you try with `allow_other` and `local` on the two systems where `bSupports2TBFiles` is not set? This makes sure FUSE does not deny file system requests and `coreservicesd` asks for the volume's capabilities.

**Comment by swt2c** [comment]
Try `allow_other` and `local` at the same time, or one by one?

**Comment by bfleischer** [comment]
At the same time.

**Comment by swt2c** [comment]
allow_other, local => bSupports2TBFiles SET
allow_other => bSupports2TBFiles NOT SET
local => bSupports2TBFiles SET

**Comment by bfleischer** [comment]
When `local` is set, coreservicesd calls `vfsop_getattr()` and caches the volume's capabilities. If you mount the volume without `local` coreservicesd does not request the volume's capabilities.

I believe the problem with non-local volumes is that a process, let's call it process A, might ask for the volume's capabilities before you call `FSGetVolumeParms()`. In this case coreservicesd caches the response A receives. If A receives bad data coreservicesd will not cache `VOL_CAP_FMT_2TB_FILESIZE` and as a consequence subsequent `FSGetVolumeParms()` calls will not return `bSupports2TBFiles`. This is something Apple needs to address.

For now the only workaround seems to be mounting the volume with the `local` option. This way coreservicesd will ask for the volume's capabilities and cache the result.

**Comment by bfleischer** [comment]
It would be interesting to find out the real identity of process A. It is most likely an application or service that is only installed on those two systems.

**Comment by swt2c** [comment]
What do you mean by "bad data?"

Any ideas on how we could figure out what process A is?

**Comment by swt2c** [comment]
Okay, I think I've identified 'process A.'  It appears that it is Adobe Creative Cloud.  When that is installed, I can now reproduce this on a third machine.  Also, if I close it on one of the original machines, the problem goes away.

**Comment by swt2c** [comment]
Okay, so after a bit of further discussion with Jim and investigation on my system, it appears that your theory is correct.  There is a process 'Adobe CEF Helper' that is part of Adobe Creative Cloud.  This process apparently requests volume capabilities upon mount - however, this process is heavily sandboxed and its requests get denied with EPERM.  This response then gets cached in coreservicesd, which leads to the problem I'm seeing where bSupports2TBFiles is not set.

Jim has fixed the bug in coreservicesd, but apparently it probably won't be released until 10.13.  So I guess I'm stuck with `-o local` for now.

Thanks for all your help on this!  I'm going to close this as there isn't anything that can be done in osxfuse about this, as far as I know.

**Comment by bfleischer** [comment]
Thanks for the update. I agree, `-o local` seems to be the only workaround until Apple releases the bug fix.

**Comment by swt2c** [comment]
As a note, it does appear that this bug has been fixed in 10.13.
